CC/include/Parallel.h

```diff
@@ -15,31 +15,50 @@
 //#                                                                        #
 //##########################################################################
 
-#ifndef PARALLEL_SORT_HEADER
-#define PARALLEL_SORT_HEADER
-
-#ifdef ParallelSort
-#undef ParallelSort
-#pragma message "Replacing preprocessor symbol 'ParallelSort' with the one defined in Parallel.h"
-#endif
+#ifndef CCCORE_PARALLEL_HEADER
+#define CCCORE_PARALLEL_HEADER
 
 #if defined(_MSC_VER) && (_MSC_VER >= 1800)
 
-	//Parallel Patterns Library (for parallel sort)
-	#include <ppl.h>
-
-	#define ParallelSort Concurrency::parallel_sort
+#define USE_PARALLEL
+//Parallel Patterns Library (for parallel sort)
+#include <ppl.h>
+#define CCParallelSort Concurrency::parallel_sort
+#define CCParallelFor Concurrency::parallel_for
+#define CCParallelForEach Concurrency::parallel_for_each
+#define CCCriticalSection Concurrency::critical_section
+//TODO: CCParallelWithLimitedThreads
 
 #elif USE_TBB
 
-	#include <tbb/parallel_sort.h>
-
-	#define ParallelSort tbb::parallel_sort
+#define USE_PARALLEL
+/*
+ * TODO: Task arena was in preview state prior to TBB 4.3 release.
+ * It's very unlikely that end user will compile with 4.2 but
+ * this version is notably used by ubuntu 14.04 on Travis. */
+#define TBB_PREVIEW_TASK_ARENA 1
+#include <tbb/parallel_for.h>
+#include <tbb/parallel_for_each.h>
+#include <tbb/parallel_sort.h>
+#include <tbb/critical_section.h>
+#include <tbb/task_arena.h>
+#include <tbb/task_group.h>
+#define CCParallelSort tbb::parallel_sort
+#define CCParallelFor tbb::parallel_for
+#define CCParallelForEach tbb::parallel_for_each
+#define CCCriticalSection tbb::critical_section
+//Limit the number of thread of a parallel template function (or task)
+#define CCParallelWithLimitedThreads(task, num_threads) { \
```

Comment:
@dgirardeau I'm nearly done with that. As I do not have Windows, could you please check how to limit concurrency with PPL? With TBB I use task_arena but there is no such feature in PPL.